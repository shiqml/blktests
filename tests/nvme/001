#!/bin/bash
#
# Regression test for commit 3d030e41d96f ("nvme: add tracepoint for
# nvme_setup_cmd")
#
# Copyright (C) 2018 Johannes Thumshirn
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

DESCRIPTION="enable nvme_setup_nvm_cmd tracepoint and submit I/O to the device"

QUICK=1

requires() {
	_have_tracefs && _have_tracepoint "nvme/nvme_setup_nvm_cmd"
}

filter_trace() {
	sed -r -e  's/dd-[0-9]+\s+/dd-XXX   /' \
		-e 's/\[[0-9]+\]/[XXX]/' \
		-e 's/[0-9\.]{4}/..../' \
		-e 's/\s+[0-9]+\.[0-9]+/     X.XXXXXX/' \
		-e 's/qid=[0-9-]+/qid=X/' \
		-e 's/nsid=[0-9-]+/nsid=X/' \
		-e 's/cmdid=[0-9-]+/cmdid=X/' \
		-e 's/slba=[0-9]+/slba=X/'
}

test_device() {
	local tfspath=/sys/kernel/debug/tracing
	local tfsinstance=${tfspath}/instances/blktests

	echo "Running ${TEST_NAME}"

	mkdir "${tfsinstance}"

	echo 1 > "${tfsinstance}/events/nvme/nvme_setup_nvm_cmd/enable"

	dd if="${TEST_DEV}" of=/dev/null iflag=direct bs=512 count=1 status=none

	filter_trace < "${tfsinstance}/trace"

	echo 0 > "${tfsinstance}/events/nvme/nvme_setup_nvm_cmd/enable"

	rmdir "${tfsinstance}"

	echo "Test complete"
}
