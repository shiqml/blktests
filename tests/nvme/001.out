Running nvme/001
# tracer: nop
#
#                              _-----=> irqs-off
#                             / _----=> need-resched
#                            | / _---=> hardirq/softirq
#                            || / _--=> preempt-depth
#                            ||| /     delay
#           TASK-PID   CPU#  ||||    TIMESTAMP  FUNCTION
#              | |       |   ||||       |         |
              dd-XXX   [XXX] ....     X.XXXXXX: nvme_setup_nvm_cmd: qid=X, nsid=X, cmdid=X, flags=0x0, meta=0x0, cmd=(nvme_cmd_read slba=X, len=0, ctrl=0x0, dsmgmt=0, reftag=0)
Test complete
